stan/math/prim/fun/ordered_constrain.hpp

```diff
@@ -14,27 +14,27 @@ namespace math {
  * free vector.  The returned constrained vector will have the
  * same dimensionality as the specified free vector.
  *
- * @tparam T type of elements in the vector
+ * @tparam T type of the vector
  * @param x Free vector of scalars.
  * @return Positive, increasing ordered vector.
  * @tparam T Type of scalar.
  */
-template <typename T>
-Eigen::Matrix<T, Eigen::Dynamic, 1> ordered_constrain(
-    const Eigen::Matrix<T, Eigen::Dynamic, 1>& x) {
+template <typename EigVec, require_eigen_col_vector_t<EigVec>* = nullptr>
+Eigen::Matrix<value_type_t<EigVec>, Eigen::Dynamic, 1> ordered_constrain(
+    const EigVec& x) {
   using Eigen::Dynamic;
   using Eigen::Matrix;
   using std::exp;
-  using size_type = index_type_t<Matrix<T, Dynamic, 1>>;
+  using size_type = Eigen::Index;
 
   size_type k = x.size();
-  Matrix<T, Dynamic, 1> y(k);
+  Eigen::Matrix<value_type_t<EigVec>, Eigen::Dynamic, 1> y(k);
```

Comment:
Could you just use the plainobject here
```suggestion
    typename EigVec::PlainObject y(k);
```

Reply:
Though I only think this matters if we wanted to support fixed size matrices

Reply:
It is less code, which is good enough for me.